File: src/main/java/org/sowatt/sowattweb/service/EnoceanDeviceListener.java
package org.sowatt.sowattweb.service;

import com.sun.org.apache.xpath.internal.operations.Bool;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.sowatt.sowattweb.domain.ButtonState;
import org.sowatt.sowattweb.domain.Switch2RockerButtonPosition;
import org.sowatt.sowattweb.integration.KNXProcessCommunicationWrapper;
import org.sowatt.sowattweb.repository.GoogleSpreadSheetDatabase;
import org.sowatt.sowattweb.web.rest.StateDataPointsResource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;
import tuwien.auto.calimero.KNXException;
import tuwien.auto.calimero.datapoint.Datapoint;
import uk.co._4ng.enocean.communication.Connection;
import uk.co._4ng.enocean.communication.DeviceListener;
import uk.co._4ng.enocean.communication.DeviceValueListener;
import uk.co._4ng.enocean.devices.DeviceManager;
import uk.co._4ng.enocean.devices.EnOceanDevice;
import uk.co._4ng.enocean.eep.EEP;
import uk.co._4ng.enocean.eep.EEPAttribute;
import uk.co._4ng.enocean.eep.EEPAttributeChangeJob;
import uk.co._4ng.enocean.eep.eep26.attributes.EEP26RockerSwitch2RockerAction;
import uk.co._4ng.enocean.eep.eep26.attributes.EEP26RockerSwitch2RockerButtonCount;
import uk.co._4ng.enocean.link.LinkLayer;

import javax.annotation.PostConstruct;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class EnoceanDeviceListener implements DeviceListener, DeviceValueListener {

    private DeviceManager deviceManager;
    private Connection connection;
    private LinkLayer linkLayer;
    @Autowired
    private GoogleSpreadSheetDatabase googleSpreadSheetDatabase;
    @Autowired
    private StateDataPointsResource stateDataPointsResource;
    @Autowired
    private KNXProcessCommunicationWrapper knxProcessCommunicationWrapper;


    private static final Logger logger = LoggerFactory.getLogger(EnoceanDeviceListener.class);

    @PostConstruct
    private void initEnoceanDevices() {
        try {
            // create the lowest link layer
            linkLayer = new LinkLayer("ttyUSB0");
            List<String> ports = LinkLayer.getCommsPorts();

//            LinkLayer linkLayer = new LinkLayer("toto");$
            for (String port : LinkLayer.getCommsPorts()
            ) {
                logger.info("Found port: {}", port);
            }
            deviceManager = new DeviceManager();

            // create a device listener for handling device updates

            deviceManager.addDeviceListener(this);
            deviceManager.addDeviceValueListener(this);

            // register a rocker switch
            for (EnOceanDevice enOceanDevice : googleSpreadSheetDatabase.getAllRockerSwitch()
            ) {
                deviceManager.registerDevice(enOceanDevice);
            }

            // create the connection layer
            connection = new Connection(linkLayer, deviceManager);

            // connect the link
            linkLayer.connect();
        } catch (Exception e) {
            System.err.println("The given port does not exist or no device is plugged in" + e);
        }
    }

    @Override
    public void addedEnOceanDevice(EnOceanDevice device) {
        logger.info("Added device: {} ({})", device.getAddressHex(), device.getEEP().getIdentifier());
    }

    @Override
    public void modifiedEnOceanDevice(EnOceanDevice device) {
        logger.info("Modified device: {} ({})", device.getAddressHex(), device.getEEP().getIdentifier());
    }

    @Override
    public void removedEnOceanDevice(EnOceanDevice device) {
        logger.info("Removed device: {} ({})", device.getAddressHex(), device.getEEP().getIdentifier());
    }

    @Override
    public void deviceAttributeChange(EEPAttributeChangeJob eepAttributeChangeJob) {
        EnOceanDevice enOceanDevice = eepAttributeChangeJob.getDevice();

        for (EEPAttribute attr : eepAttributeChangeJob.getChangedAttributes()) {
            for (Switch2RockerButtonPosition buttonPosition : Switch2RockerButtonPosition.values()
            ) {
                ButtonState buttonState = googleSpreadSheetDatabase.findButtonStateBy(enOceanDevice, buttonPosition);
                if (attr instanceof EEP26RockerSwitch2RockerAction) {
                    Boolean buttonValue = ((EEP26RockerSwitch2RockerAction) attr).getButtonValue(buttonPosition.ordinal());
                    buttonState.setPressed(buttonValue);
                    if (buttonValue) {
                        buttonState.setDatePressed(LocalDateTime.now());
                    }
                }

                if (attr instanceof EEP26RockerSwitch2RockerButtonCount) {
                    EEP26RockerSwitch2RockerButtonCount buttonCount = (EEP26RockerSwitch2RockerButtonCount) attr;
                    if (buttonState.getPressed() && (buttonCount.getValue() == 0)) {
                        assert buttonState.getDatapoints().size() == 1;
                        Datapoint dp = buttonState.getDatapoints().get(0);
                        try {
                            Boolean previousValue = knxProcessCommunicationWrapper.readBool(dp.getMainAddress());
                            knxProcessCommunicationWrapper.write(dp.getMainAddress(), !previousValue);
                        } catch (KNXException e) {
                            e.printStackTrace();
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }

                    }
                }
            }
            logger.info("Device: {} Channel: {} Attribute: {} Value: {}", eepAttributeChangeJob.getDevice().getAddressHex(), eepAttributeChangeJob.getChannelId(), attr.getName(), attr.getValue());
        }
    }
}

File: src/main/kotlin/org/sowatt/sowattweb/integration/EnoceanDeviceListener.kt
package org.sowatt.sowattweb.integration

import org.slf4j.LoggerFactory
import org.sowatt.sowattweb.domain.Button
import org.sowatt.sowattweb.domain.ControlPoint
import org.sowatt.sowattweb.domain.types.Switch2RockerButtonPosition
import org.sowatt.sowattweb.repository.ButtonRepository
import org.sowatt.sowattweb.repository.ControlPointRepository
import org.sowatt.sowattweb.repository.ToggleCommandRepository
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import uk.co._4ng.enocean.communication.Connection
import uk.co._4ng.enocean.communication.DeviceListener
import uk.co._4ng.enocean.communication.DeviceValueListener
import uk.co._4ng.enocean.devices.DeviceManager
import uk.co._4ng.enocean.devices.EnOceanDevice
import uk.co._4ng.enocean.eep.EEPAttributeChangeJob
import uk.co._4ng.enocean.eep.eep26.attributes.EEP26RockerSwitch2RockerAction
import uk.co._4ng.enocean.eep.eep26.attributes.EEP26RockerSwitch2RockerButtonCount
import uk.co._4ng.enocean.link.LinkLayer
import javax.annotation.PostConstruct
import javax.persistence.EntityManager


@Component
public class EnoceanDeviceListener(private val knxProcessCommunicationWrapper: KNXProcessCommunicationWrapper, private val toggleCommandRepository: ToggleCommandRepository, private val deviceRepository: ControlPointRepository, private val buttonRepository: ButtonRepository, val entityManager: EntityManager): DeviceListener, DeviceValueListener {
    private val logger = LoggerFactory.getLogger(EnoceanDeviceListener::class.java)

    lateinit private var linkLayer: LinkLayer
    @Value("\${serial.port.id:ttyUSB0}")
    lateinit private var serialPortId: String
    private var deviceManager: DeviceManager= DeviceManager()
    lateinit private var connection: Connection


    @PostConstruct
    private fun initEnoceanDevices() {
        try {
            // create the lowest link layer
            linkLayer = LinkLayer(serialPortId);

            //In case configured port is incorrect
            for (port: String  in LinkLayer.getCommsPorts()
            ) {
                logger.info("Found port: {}", port);
            }

            // create a device listener for handling device updates

            deviceManager.addDeviceListener(this);
            deviceManager.addDeviceValueListener(this);

            // register a rocker switch
            for (device: ControlPoint in deviceRepository.findAll())
                deviceManager.registerDevice(device.toEnoceanDevice());

            // create the connection layer
            connection = Connection(linkLayer, deviceManager);

            // connect the link
            linkLayer.connect();
        } catch ( e: Exception) {
            System.err.println("The given port does not exist or no device is plugged in" + e);
        }
    }

    override fun modifiedEnOceanDevice(changedDevice: EnOceanDevice?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun removedEnOceanDevice(changedDevice: EnOceanDevice?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun addedEnOceanDevice(device: EnOceanDevice?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deviceAttributeChange(eepAttributeChangeJob: EEPAttributeChangeJob) {
        val enOceanDevice = eepAttributeChangeJob.device
        for (changedAttribute in eepAttributeChangeJob.changedAttributes) {
            loop@ for (buttonPosition: Switch2RockerButtonPosition in Switch2RockerButtonPosition.values()) {
                val button: Button = buttonRepository.getButton(enOceanDevice.addressHex, buttonPosition)
                        ?: continue
                when  (changedAttribute) {
                    is EEP26RockerSwitch2RockerAction -> {
                        val buttonValue = changedAttribute.getButtonValue(buttonPosition.ordinal) //TODO ugly trick: the enum keeps the ordering of the standard...
                        button.isPressed = buttonValue
                    }
                    is EEP26RockerSwitch2RockerButtonCount  -> {
                        if (button.isPressed && changedAttribute.value == 0) {//TODO how can we put test condition on one line
                            val toggleCommand = toggleCommandRepository.findByButton(button)?: continue@loop
                            assert(toggleCommand.switchList.size == 1)

                            val switch = toggleCommand.switchList[0]
                            try {
                                val previousValue: Boolean = knxProcessCommunicationWrapper.readBool(switch.toCommandDP().mainAddress)
                                knxProcessCommunicationWrapper.write(switch.toCommandDP().mainAddress, !previousValue)
                            } catch (e: java.lang.Exception) {
                                logger.error("Error while updating switch ${switch.groupAddress}",e)
                            }
                            button.isPressed=false
                        }
                    }
                }
                entityManager.persist(button)

            }
            logger.info("Device: {} Channel: {} Attribute: {} Value: {}", eepAttributeChangeJob.device.addressHex, eepAttributeChangeJob.channelId, changedAttribute.name, changedAttribute.value)
        }
    }

}